### PR TITLE
[Rust] Use CoW for escaping JSON Pointers

### PR DESCRIPTION
**Pull request by juharris**
Switch JSON pointer escaping to use `Cow<str>` in order to avoid unnecessary allocations and support both borrowed and owned string inputs.

Loading benchmark: 1-2% faster.

Provider: Add tests for escaping keys

Fix bug: very specific Configurable String edge case: did not handle Configurable Strings if the top-level (first) key had "~" or "/"s.

### CODE REVIEW

File: rust/optify/src/json/escape_json_pointer.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+macro_rules! escape_json_pointer {
```
Comment (tavianator) [inline rust/optify/src/json/escape_json_pointer.rs:1]: Does this need to be a macro?  I think you could use a normal function:

```rust
fn escape_json_pointer(s: &str) -> Cow<'_, str> {
  s.cow_replace("~", "~0").cow_replace("/", "~1")
}
```

Comment (juharris) [reply rust/optify/src/json/escape_json_pointer.rs:1]: I tried stuff like that, but you can't chain `cow_replace` because of ownership. You gotta assign, then ownership gets awkward. I was able to make it a function, but it was pretty convoluted with nested `match`s.

Comment (juharris) [reply rust/optify/src/json/escape_json_pointer.rs:1]: Discussed a bit and it's tricky. Happy to change it later.
